Improve BigInteger casting behaviours

vexx32 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Added a few special cases and specific casting methods for BigInteger in LanguagePrimitives' casting methods and cached conversions.

Specific conversion paths for boolean & nulls were added, and conversions added to the cache. Also, some existing conversion paths were modified to add special cases for BigInteger, which are needed because it's not `IConvertible` like other numeric types are.

Fixes #12623

## PR Context

These changes were split from #11634 and will fix #12623. 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

vexx32 committed:
:white_check_mark: Add conversion tests

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -2908,27 +2909,45 @@ private static bool TryScanNumber(string strToConvert, Type resultType, out obje
             {
                 typeConversion.WriteLine("Returning numeric zero.");
+
+                // BigInteger is not IConvertible and will throw from ChangeType.
```
My ask https://github.com/dotnet/runtime/issues/36264

vexx32 replied:
I'm curious as to their response, but I think it's unlikely it'll change overall. If `BigInteger` gets to be `IConvertible`, sooner or later someone's gonna come along and ask if `System.Numerics.Complex` should be as well, and... let's just say I'm watching that debate from the sidelines. 😂 🍿

PoshChan commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -14,6 +14,7 @@
 using System.Management.Automation.Internal;
```
@vexx32, your last commit had 2 failures in `PowerShell-CI-linux`
method conversion.Correctly casts 100n to value 100 as type [int]
```powershell
Expected '100', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```
method conversion.Correctly casts 1234s to value 1234 as type [bigint]
```powershell
Expected '1234', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-macos`
method conversion.Correctly casts 100n to value 100 as type [int]
```powershell
Expected '100', but got $null.
at <ScriptBlock>, /Users/runner/runners/2.168.2/work/1/s/test/powershell/Language/Parser/Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```
method conversion.Correctly casts 1234s to value 1234 as type [bigint]
```powershell
Expected '1234', but got $null.
at <ScriptBlock>, /Users/runner/runners/2.168.2/work/1/s/test/powershell/Language/Parser/Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-windows`
method conversion.Correctly casts 100n to value 100 as type [int]
```powershell
Expected '100', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```
method conversion.Correctly casts 1234s to value 1234 as type [bigint]
```powershell
Expected '1234', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
method conversion.Correctly casts 100n to value 100 as type [int]
```powershell
Expected '100', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-macos`
method conversion.Correctly casts 100n to value 100 as type [int]
```powershell
Expected '100', but got $null.
at <ScriptBlock>, /Users/runner/runners/2.168.2/work/1/s/test/powershell/Language/Parser/Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-windows`
method conversion.Correctly casts 100n to value 100 as type [int]
```powershell
Expected '100', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```

PoshChan replied:
@vexx32, your last commit had 6 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimInstance -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimInstance -Namespace root/Int*ro' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'New-CimInstance -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Invoke-CimMethod -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimClass -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
method conversion.Correctly casts 100n to value 100 as type [int]
```powershell
Expected '100', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-windows`
Validate 'Save-Help -DestinationPath for all PowerShell modules..Validate Save-Help for the 'Microsoft.PowerShell.Diagnostics' module
```powershell
Expected 'Microsoft.PowerShell.Diagnostics_ca046f10-ca64-4740-8ff9-2565dba61a4f_en-US_helpcontent.cab', because Save-Help for  should download 'Microsoft.PowerShell.Diagnostics_ca046f10-ca64-4740-8ff9-2565dba61a4f_en-US_helpcontent.cab', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Help\UpdatableHelpSystem.Tests.ps1: line 333
333:     $compressedFile | Should -Be $expectedCompressedFile -Because "Save-Help for $module should download '$expectedCompressedFile'"
```
method conversion.Correctly casts 100n to value 100 as type [int]
```powershell
Expected '100', but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-macos`
method conversion.Correctly casts 100n to value 100 as type [int]
```powershell
Expected '100', but got $null.
at <ScriptBlock>, /Users/runner/runners/2.168.2/work/1/s/test/powershell/Language/Parser/Conversions.Tests.ps1: line 510
510:         $Result | Should -Be $Value
```

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -445,7 +445,8 @@ private static void InitializeGetEnumerableCache()
         {
-            if (type == null) { return false; }
+            if (type == null)
+            { return false; }
```
```suggestion
            {
                return false;
            }
```

vexx32 replied:
Argh, autoformatter got me again. I'll do a style pass after this PR, but I'll take these out for now as they aren't functional changes.

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -473,7 +474,8 @@ public static bool IsObjectEnumerable(object obj)
             obj = PSObject.Base(obj);
-            if (obj == null) { return null; }
+            if (obj == null)
+            { return null; }
```
```suggestion
            {
                return null;
            }
```

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -664,7 +666,8 @@ public static bool Equals(object first, object second, bool ignoreCase, IFormatP
             }
 
-            if (first.Equals(second)) return true;
+            if (first.Equals(second))
+                return true;
```
```suggestion
            if (first.Equals(second))
            {
                return true;
            }
```

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -1051,7 +1062,8 @@ internal static bool IsTrue(IList objectArray)
                     }
 
-                    if (firstElement.Count < 1) return false;
+                    if (firstElement.Count < 1)
+                        return false;
```
```suggestion
                    if (firstElement.Count < 1)
                    {
                        return false;
                    }
```

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -2886,8 +2910,16 @@ private static bool TryScanNumber(string strToConvert, Type resultType, out obje
             try
             {
+                var parsedNumber = Parser.ScanNumber(strToConvert, resultType, shouldTryCoercion: false);
+                if (resultType == typeof(BigInteger) || parsedNumber is BigInteger)
```
What if we request `Int` (resultType) but get `BigInteger` in `parsedNumber`? Should we return `false`?

vexx32 replied:
No. `[int]"20s"` parses correctly (we get `short` for `parsedNumber` but the `resultType` desired there is `int`), so too should things like `[long]"10n"` (`BigInteger` for `parsedNumber` but `resultType` of `long`)

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -2900,36 +2932,55 @@ private static bool TryScanNumber(string strToConvert, Type resultType, out obje
-            TypeConverter integerConverter = LanguagePrimitives.GetIntegerSystemConverter(resultType);
+
+            // This will contain null if resultType == typeof(BigInteger)
+            TypeConverter integerConverter = GetIntegerSystemConverter(resultType);
             try
             {
                 if (TryScanNumber(strToConvert, resultType, out object result))
                 {
                     return result;
                 }
-                else
+
+                if (resultType == typeof(BigInteger))
```
Why do we need `integerConverter` if we can return after TryScanNumber?
And why do we check `resultType == typeof(BigInteger)` if `integerConverter` could be null already?
I think we should refactor the code.

vexx32 replied:
I couldn't see a good way to refactor that without completely diverging the code path, but I'll take another look.

vexx32 replied:
Looks like we don't even need the variable, you're right. I'll redo this a bit more.

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -4623,7 +4706,8 @@ internal static PSObject SetObjectProperties(object o, IDictionary properties, T
                                     if (TypeResolver.TryResolveType(property.TypeNameOfValue, out propType))
                                     {
-                                        if (formatProvider == null) { formatProvider = CultureInfo.InvariantCulture; }
+                                        if (formatProvider == null)
+                                        { formatProvider = CultureInfo.InvariantCulture; }
```
```suggestion
                                        if (formatProvider == null)
                                        {
                                            formatProvider = CultureInfo.InvariantCulture;
                                        }
```

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -5025,7 +5109,8 @@ internal static object ThrowInvalidConversionException(object valueToConvert, Ty
                     var signatureType = signatureEnumerator.Current;
                     // Skip the non-bindable signatures
-                    if (signatureType == typeof(Func<PSNonBindableType>)) { continue; }
+                    if (signatureType == typeof(Func<PSNonBindableType>))
+                    { continue; }
```
```suggestion
                    if (signatureType == typeof(Func<PSNonBindableType>))
                    {
                        continue;
                    }
```

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -5038,7 +5123,8 @@ internal static object ThrowInvalidConversionException(object valueToConvert, Ty
 
                         // If there is no exact match, then we use the first compatible signature we found.
-                        if (matchedIndex == -1) { matchedIndex = index; }
+                        if (matchedIndex == -1)
+                        { matchedIndex = index; }
```
```suggestion
                        if (matchedIndex == -1)
                        {
                            matchedIndex = index;
                        }
```

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -5088,7 +5174,8 @@ internal bool ProjectedSignatureMatchesTarget(Type[] argumentTypes, out bool sig
                 signaturesMatchExactly = false;
                 int length = argumentTypes.Length;
-                if (length != targetParameters.Length + 1) { return false; }
+                if (length != targetParameters.Length + 1)
+                { return false; }
```
```suggestion
                if (length != targetParameters.Length + 1)
                {
                    return false;
                }
```

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -5122,7 +5209,8 @@ private static bool ProjectedTypeMatchesTargetType(Type sourceType, Type targetT
                 if (targetType.IsByRef || targetType.IsPointer)
                 {
-                    if (!sourceType.IsGenericType) { return false; }
+                    if (!sourceType.IsGenericType)
+                    { return false; }
```
```suggestion
                    if (!sourceType.IsGenericType)
                    {
                        return false;
                    }
```

vexx32 committed:
:bug: Improve biginteger casting behaviours

Added a few special cases and specific casting methods for BigInteger
in LanguagePrimitives' casting methods and cached conversions.

Fixes #12623

iSazonov commented:
@daxian-dbw Please review.

iSazonov commented:
@daxian-dbw Could you please review?

vexx32 commented:
@daxian-dbw friendly ping for review when you have a moment. Thanks! 💖

iSazonov commented:
Reopen to update CIs.

vexx32 commented:
I probably need to rebase this... will have a look this afternoon :)

daxian-dbw commented:
I apologize for not being able to review this in time. There are a few other @vexx32's and @iSazonov's PRs that we intended to get to but just couldn't yet ☹️ I will get to this one today.

daxian-dbw commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -4132,14 +4178,22 @@ private class ConvertCheckingForCustomConverter
+            // Handle BigInteger first, as it is not IConvertible
+            if (resultType == typeof(BigInteger))
+            {
+                return default(BigInteger);
```
I guess `default(BigInteger) == BigInteger.Zero`, right? It might be more clear to use `BigInteger.Zero` here.

vexx32 replied:
```suggestion
                return BigInteger.Zero;
```

daxian-dbw reviewed (approved):
LGTM with a minor comment.

vexx32 committed:
:recycle: Address review comment

iSazonov commented on src/System.Management.Automation/engine/LanguagePrimitives.cs:
```diff
@@ -3220,6 +3248,24 @@ private static bool TryScanNumber(string strToConvert, Type resultType, out obje
+            PSObject originalValueToConvert,
+            IFormatProvider formatProvider,
+            TypeTable backupTable)
+                => ((BigInteger)valueToConvert) != default(BigInteger);
```
Maybe follow the same pattern?
```suggestion
                => ((BigInteger)valueToConvert) != BigInteger.Zero;
```

vexx32 committed:
:recycle: Accept code review suggestion

Co-authored-by: Ilya <darpa@yandex.ru>

daxian-dbw reviewed (approved):
Perfect, thanks! I appreciate your patience!

iSazonov commented:
😺 Could add cognac labels :wine_glass: :
- if PR waits one month - one start
- if PR waits 3 months - two starts
- and so on

daxian-dbw commented:
>if PR waits one month - one start
>if PR waits 3 months - two starts

Or use angry face 😠 instead

iSazonov commented:
> Or use angry face 😠 instead

It is already reserved for user's Issues about bugs. :-)
